### PR TITLE
EZP-27101: Remove recursion when retrieving search metadata for relation fields

### PR DESCRIPTION
**Pull request by natanael89**
JIRA issue: [EZP-27101](https://jira.ez.no/browse/EZP-27101)

When retrieving the search metadata from objects containing "Object relation" and "Object relations" field type, there is a recursion protection in place that ensures that there is no infinite recursion caused by circular references. However, it only checks if the certain object was already processed and this is causing problems when the same object is referenced from two different fields (as described in [EZP-27101](https://jira.ez.no/browse/EZP-27101)). 

This PR removes the recursion completely by making the relations load metadata only one level deep. By removing recursion, I was also able to remove the recursion protection, which solves the original issue.

This PR introduces a slight BC break because it changes logic (previously the metadata was loaded recursively and now it is loaded only one level deep), but the new behavior allows for better (more relevant) search results.
An alternative solution (without a BC break, but which keeps the broken logic of loading the metadata recursively) can be found here: https://github.com/ezsystems/ezpublish-legacy/pull/1371.

### CODE REVIEW

**Comment by andrerom** [comment]
@natanael89 for the BC break, can you add description of that in doc/bc/5.90/relation_indexing.md ?

This will be for the 2018.06 release.

File: kernel/classes/ezcontentobjectattribute.php [hunk 1]
```diff
@@ -1168,26 +1168,34 @@ function fromString( $string )
...
-            if ( $classAttribute->attribute( 'is_searchable' ) )
+            if ( !$classAttribute->attribute( 'is_searchable' ) )
             {
-                $attributeMetaData = $attribute->metaData();
-                if ( $attributeMetaData !== false )
+                continue;
+            }
+
+            $datatype = $attribute->dataType();
+            if($skipRelationAttributes && $datatype->isRelationType())
```
Comment (andrerom) [inline kernel/classes/ezcontentobjectattribute.php]: legacy cs: `if( $skipRelationAttributes && $datatype->isRelationType() )`

File: kernel/search/plugins/ezsearchengine/ezsearchengine.php [hunk 2]
```diff
@@ -68,7 +68,6 @@ public function addObject( $contentObject, $commit = true )
         $placement = 0;
         $previousWord = '';
 
-        eZContentObject::recursionProtectionStart();
```
Comment (andrerom) [inline kernel/search/plugins/ezsearchengine/ezsearchengine.php]: wondering if we might need to keep these in, in case custom relation field types stiull use it.

or will it cause problems?

Comment (natanael89) [reply kernel/search/plugins/ezsearchengine/ezsearchengine.php]: We can (and should) keep those, you are right.

**Comment by andrerom** [comment]
@pkamps Is this more in line with what you suggested i other other PR? Do you approve? :)

@gggeek Any opinions on this in terms of search use?

**Comment by natanael89** [comment]
I added the BC description and made the suggested fixes.

File: kernel/classes/ezcontentobjectattribute.php [hunk 3]
```diff
@@ -1182,7 +1182,7 @@ static function metaDataArray( &$attributes, $skipRelationAttributes = false )
             }
 
             $datatype = $attribute->dataType();
-            if($skipRelationAttributes && $datatype->isRelationType())
+            if( $skipRelationAttributes && $datatype->isRelationType() )
```
Comment (andrerom) [inline kernel/classes/ezcontentobjectattribute.php]: my prev comment forgot to also specify whitespace after `if`

File: doc/bc/5.90/relation_indexing.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# Relation search indexing changes 
+
+For the 2018.06 release, eZ Publish changed the way relations are indexed. Previously the relation metadata was loaded recursively, meaning that relations of relations etc. were also included. This behavior was changed to only include Objects one level deep so only direct relations are taken into account.
```
Comment (andrerom) [inline doc/bc/5.90/relation_indexing.md]: should explain why the change was done.

like `are taken into account, in order to solve various recursion problems when relations are circular.`

File: kernel/classes/ezcontentobjectattribute.php (hunk 3 above)
Comment (natanael89) [reply kernel/classes/ezcontentobjectattribute.php]: Fixed.

File: doc/bc/5.90/relation_indexing.md (hunk 4 above)
Comment (natanael89) [reply doc/bc/5.90/relation_indexing.md]: I added another sentence for the "why" part.
